<style> html {font-size:12} </style>

In [1]:
from IPython.display import display
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import pandas as pd

# 1.- Regresión Lineal ordinaria
<span style="font-size:12pt"> a)

In [2]:
df = pd.read_csv("kc_house_data.csv")
df.drop(['id','date','zipcode'],axis=1,inplace=True)
df.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0  221900.0         3       1.00         1180      5650     1.0           0   
1  538000.0         3       2.25         2570      7242     2.0           0   
2  180000.0         2       1.00          770     10000     1.0           0   
3  604000.0         4       3.00         1960      5000     1.0           0   
4  510000.0         3       2.00         1680      8080     1.0           0   

   view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0     0          3      7        1180              0      1955             0   
1     0          3      7        2170            400      1951          1991   
2     0          3      6         770              0      1933             0   
3     0          5      7        1050            910      1965             0   
4     0          3      8        1680              0      1987             0   

       lat     long  sqft_living15  sqft_lot15  
0  47.5112 -122.257           1340        5650  
1  47.7210 -122.319           1690        7639  
2  47.7379 -122.233           2720        8062  
3  47.5208 -122.393           1360        5000  
4  47.6168 -122.045           1800        7503

<span style="font-size:12pt"> Debido a que lo que se quiere estudiar es el precio de las casas los atributos de id, fecha y código postal son irrelevantes, entonces se eliminan estas caracteristicas del dataset con la línea 4.
Notar que la fecha y ubicación puede influr, pero en este caso todas las casa están en la misma locación y se venden en un corto periodo de tiempo (1 año).</span>

<span style="font-size:12pt"> b)

In [3]:
df.shape
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 18 columns):
price            21613 non-null float64
bedrooms         21613 non-null int64
bathrooms        21613 non-null float64
sqft_living      21613 non-null int64
sqft_lot         21613 non-null int64
floors           21613 non-null float64
waterfront       21613 non-null int64
view             21613 non-null int64
condition        21613 non-null int64
grade            21613 non-null int64
sqft_above       21613 non-null int64
sqft_basement    21613 non-null int64
yr_built         21613 non-null int64
yr_renovated     21613 non-null int64
lat              21613 non-null float64
long             21613 non-null float64
sqft_living15    21613 non-null int64
sqft_lot15       21613 non-null int64
dtypes: float64(5), int64(13)
memory usage: 3.0 MB


price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  2.161300e+04   
mean   5.400881e+05      3.370842      2.114757   2079.899736  1.510697e+04   
std    3.671272e+05      0.930062      0.770163    918.440897  4.142051e+04   
min    7.500000e+04      0.000000      0.000000    290.000000  5.200000e+02   
25%    3.219500e+05      3.000000      1.750000   1427.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068800e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             floors    waterfront          view     condition         grade  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean       1.494309      0.007542      0.234303      3.409430      7.656873   
std        0.539989      0.086517      0.766318      0.650743      1.175459   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above  sqft_basement      yr_built  yr_renovated           lat  \
count  21613.000000   21613.000000  21613.000000  21613.000000  21613.000000   
mean    1788.390691     291.509045   1971.005136     84.402258     47.560053   
std      828.090978     442.575043     29.373411    401.679240      0.138564   
min      290.000000       0.000000   1900.000000      0.000000     47.155900   
25%     1190.000000       0.000000   1951.000000      0.000000     47.471000   
50%     1560.000000       0.000000   1975.000000      0.000000     47.571800   
75%     2210.000000     560.000000   1997.000000      0.000000     47.678000   
max     9410.000000    4820.000000   2015.000000   2015.000000     47.777600   

               long  sqft_living15     sqft_lot15  
count  21613.000000   21613.000000   21613.000000  
mean    -122.213896    1986.552492   12768.455652  
std        0.140828     685.391304   27304.179631  
min     -122.519000     399.000000     651.000000  
25%     -122.328000    1490.000000    5100.000000  
50%     -122.230000    1840.000000    7620.000000  
75%     -122.125000    2360.000000   10083.000000  
max     -121.315000    6210.000000  871200.000000

<span style="font-size:12pt">Son 21613 entradas de datos, estas entradas tienen 18 atributos (precio, dormitorios, baños, etc) no nulos. Hay detalles estadisticos sobre el conjunto de muestra por atributo en la tabla que se muestra arriba con el metodo `.describe()`</span>

<span style="font-size:12pt">c) </span>

In [4]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)

"""
plt.figure(figsize = (16, 5))
plt.hist(df_scaled['price'], bins = 10)
plt.show()
"""

df_scaled['price'] = np.log(df['price'])
df_scaled.head()
df_scaled.describe()

"""
plt.figure(figsize = (16, 5))
plt.hist(df_scaled['price'], bins = 10)
plt.show()
"""

print("Forma",df_scaled.shape)

Forma (21613, 18)


<span style="font-size:12pt">
En las mediciones se midieron 18 atributos distintos, en diferentes escalas, dinero, cantidad de piezas, baños, etc. Entonces para dejar todas las cantidades en la misma escala es necesario realizar la estandarización. <br>
Al ver el histograma del precio después de estandarizar se aprecia que la gran mayoria de los datos se concetra al inicio (pegados al eje y), entonces es necesario realizar una transformación para que los datos se repartan de mejor manera a lo largo del eje. Se observa que los datos presentan un comportamiento exponencial decreciente, entonces se aplica el logaritmo para que así estos se puedan diferenciar unos de otros.<br><br>
d)
</span>

In [5]:
import sklearn.linear_model as lm
X = df_scaled.iloc[:,1:] #Gran matriz X que le quite el atributo del precio
#use .ix instead, in older pandas version
N = X.shape[0]#Cantidad de filas
X.insert(X.shape[1], 'intercept', np.ones(N))#Pega una columna de 1's
y = df_scaled['price']
#mascara estatica con el 70% de los datos
mascara = np.zeros(len(X))
limit = int(len(X)*0.7)
mascara[:limit] = 1
istrain = mascara== 1
Xtrain = X[istrain]
ytrain = y[istrain]
Xtest = X[np.logical_not(istrain)]
ytest = y[np.logical_not(istrain)]
linreg = lm.LinearRegression(fit_intercept = False)
linreg.fit(Xtrain, ytrain)

LinearRegression(copy_X=True, fit_intercept=False, n_jobs=1, normalize=False)

<span style="font-size:12pt">
El modelo de regresión lineal es
$$ y = \beta_0 + \sum_{j=1}^{p} \beta_j x_j  $$
Se puede meter el termino $\beta_0$ dentro del vector $\beta$ si es que al final de cada muestra (matriz $X$) se añade un 1, entonces al momento de resolver el sistema $X \cdot \beta = Y$, dentro del vector $\beta$ la última fila será el $\beta_0$.<br>
Lo que hace la linea 4 es exactamente añadir una columna de 1's a la matriz $X$.<br>
<br>
Los parámetros usados en la regresión lineal<br>
Para la función `lm.LinearRegression()`<br>
* fit_intercept = False. Sirve para ver si se va a calcular o no el intercepto, por defaul True<br>
* normalize = Se ignora si fit_intercept es False. Si es True la data X va a ser normalizada<br>
* copy_X = Boleano que dice si las operaciones serán o no inplace (default True)<br>
* n_jobs = los jobs usados para el procesamiento (default True)<br>

<span style="font-size:12pt">
Para el método `fit()`<br>
* X = La matriz $X$
* y = El vector y de soluciones conocidas para el input $X$
</span>

<span style="font-size:12pt">
d)

In [8]:
from numpy.linalg import inv

pesos = linreg.coef_ #el vector beta

#for i in range(len(pesos)):
#    if pesos[i]<0:
#        print("NANI!")

yEstimado = linreg.predict(Xtest)
yReal = np.array(ytest)

#Procedo a estimar la desviación estandar
std = np.sqrt(sum(map(lambda x: x**2, (yReal - yEstimado)))) / (len(yReal) - len(pesos))

#Procedo a sacar la diagonal de la pseudo matriz de covarianzas

V = inv(np.dot(np.asarray(Xtest).T , np.asarray(Xtest))).diagonal()

zScore = list()
for i in range(len(pesos)):
    zScore.append(pesos[i] / (std * np.sqrt( V[i] )))

#for i in range(len(zScore)):
#    if zScore[i]<0:
#        print("NANI!!!!")

respuesta = [("Pesos", pesos) , ("Z-Score", zScore)]
display(pd.DataFrame.from_items(respuesta))


Pesos        Z-Score
0   -0.008270    -172.144090
1    0.055960     787.124375
2    0.057769       0.000026
3    0.022159     358.830220
4    0.037595     708.243787
5    0.033526     771.298130
6    0.042082     851.132762
7    0.045277     922.818517
8    0.186212    2606.572143
9    0.047130       0.000023
10   0.031699       0.000029
11  -0.109875   -1766.945975
12   0.014798     321.406162
13   0.186277    4342.146124
14  -0.004079     -85.193885
15   0.085933    1324.630720
16  -0.007013    -118.034923
17  13.039692  314505.059680

El z-score se calcula como $$\textit{z-score} = \frac{\hat{\beta_j}}{\hat{\sigma} \cdot \sqrt(\upsilon_{jj})}$$

Donde $\hat{\beta_j}$ corresponden a los pesos del atributo j, $\hat{\sigma}$ es la desviación estandar calculada de la forma $$\hat{\sigma} = \frac{\sqrt{ \sum_{i=1}^{n} (y^{(i)} - \hat{y}^{(i)})^2}} {n - d - 1}$$

y $\upsilon_{jj}$ corresponde a la diagonal de la pseudo matriz de covarianzas (le faltaría el $\frac{1}{n}$), es decir $$\upsilon_{jj} = (X^{T} \cdot X)^{-1} _{jj}$$